docs/source/deployment/configuration.rst
.. _configuration:

=============
Configuration
=============

.. note::
  
  The majority of users will use Balena, so we'll give examples in balena here.
  Configuring in a manually installed environment is broadly the same.


Add a service account
=====================

Using the name generated by balena when the device was added, create
a dedicated service account by following :ref:`these instructions <creating_service_accounts>`.

Once you have the ``json`` file containing credentials on your computer,
select your new device from the fleet and open a terminal (on ``main``, not on the host OS) and do:

.. code-block:: shell

   nano $GOOGLE_APPLICATION_CREDENTIALS
   
Then paste the contents of the credentials file. Press ``ctrl-x`` then ``y`` to save and exit.
This file will persist over reboots of the device; you shouldn't need to touch it again.

.. figure:: /images/balena/opening-a-terminal.png
    :width: 600px
    :align: center
    :figclass: align-center

    Open a terminal to the 'main' target and you can shell into the machine directly.

.. _add_routine_and_configuration_files:

Add routine and configuration files
===================================

We'll start with an empty routine file, which you can change later (see :ref:`routine_files`), and a basic configuration file.

.. code-block:: shell

    mkdir -p /data/routines && echo "[]" > /data/routines/my-routine.json
    mkdir -p /data/configurations && nano /data/configurations/my-configuration.json

Then paste in the following JSON (a basic configuration), update the values and save it:

.. code-block:: json

   {
     "gateway": {
       "installation_reference": "my-installation-reference", // change this to a meaningful value, eg "aventa-initial-deployment"
       "latitude": 0,
       "longitude": 0,
       "receiver_firmware_version": "unknown"
     },
     "nodes": {
       "1": {},
       "2": {},
       "3": {},
       "4": {},
       "5": {}  // Remove nodes if you know you don't need them
     },
     "measurement_campaign": {
       "label": "test-campaign",
       "description": "This field can be used to label and describe different measurement campaigns"
     }
   }

.. warning::
    
    You can store as many different configurations and routines as you want, but you should always save 
    them somewhere in the ``/data/`` folder, because it persists over restarts.


Set device variables
====================

When you run the gateway, you'll be able to able to specify a routine file and a configuration file.
However, that makes for a lot of typing, especially when you're trying to debug things.

To ease frustration, the best thing to do is to set these values as environment variables.
Go to the "Device Variables" tab and add two variables, ``GATEWAY_ROUTINE_FILE`` and ``GATEWAY_CONFIG_FILE``,
whose values match the paths you set for the two files :ref:`above <add_routine_and_configuration_files>`.

.. figure:: /images/balena/device-variables.png
    :width: 600px
    :align: center
    :figclass: align-center

    Set device variables to change the default ``GATEWAY_ROUTINE_FILE`` and ``GATEWAY_CONFIG_FILE``.


Check the installation
======================

In the balena (or your own, for a manual installation) terminal, check by typing:

.. code-block:: shell
    
    gateway --help

.. figure:: /images/balena/gateway-help.png
    :width: 600px
    :align: center
    :figclass: align-center

    If the gateway is correctly installed, you should see this.

.. tip::

   You can always use the ``$GATEWAY_CONFIG_FILE`` or ``$GATEWAY_ROUTINE_FILE`` to see the paths and their contents:

   .. code-block:: shell

      echo The config file is at $GATEWAY_CONFIG_FILE and it contains...
      cat $GATEWAY_CONFIG_FILE



docs/source/usage/configuration_files.rst
.. _configuration_files:

===================
Configuration Files
===================

Configuration options for the gateway are supplied via a configuration file,
which was set up in :ref:`configuration`.

Once the ``gateway start`` command is invoked, the configuration is saved with
the output data (if saving data locally). The configuration is also added to the
metadata on the output files uploaded to the cloud, where it is used by the cloud
ingress to populate the database.

Specifying other configuration files
====================================

The easiest way of specifying the file is to set the ``GATEWAY_CONFIG_FILE`` 
environment variable.

But, there are other ways. If the environment variable is not set, ``data-gateway``
looks for a file named ``config.json`` in the working directory, or the file path
can be overridden in the CLI options (also see ``gateway start --help``):

.. code-block:: shell

    gateway start --config-file=</path/to/config.json>


Useful customisations
=====================

The most useful customisation is to add a ``measurement_campaign_reference`` field:

.. code-block:: javascript

    {
      "gateway": {
        // ...
      },
      "nodes": {
        // ...
      }
      "measurement_campaign": {
        // If you leave out this reference, a new one is created every
        // time you start the gateway. That allows you to then filter
        // down results in the dashboard to the exact run you're doing
        // right now.
        // But, if you're doing a series of related runs and want all
        // the results to be able to be merged, the set the reference
        // value here for continuity across runs. Don't forget to change
        // or remove this if you reuse the file for something else, though.
        "measurement_campaign_reference": "my-measurement-campaign",
        // You can enable further sorting of data by adding
        // campaign-specific labels
        "label": "run-1",
        // And add notes as aide-memoires.
        "description": "It's windy right now and the battery charged up overnight so we're taking the opportunity to run with mic and diff baros turned on."
      }
    }

Further customisation
=====================

Any of the options in the ``data-gateway`` `configuration module <https://github.com/aerosense-ai/data-gateway/blob/main/data_gateway/configuration.py>`_
can be customised by updating entries in the configuration file.

.. warning::
  Moving off the beaten track, especially customising things like handles and packet keys, you really have to know what you're doing!

Here is an example of a more extensive configuration file.

.. code-block:: javascript

    {
      "gateway": {
        "baudrate": 2300000,
        "endian": "little",
        "installation_reference": "my_installation_reference",
        "latitude": 0,
        "longitude": 0,
        "packet_key": 254,
        "packet_key_offset": 245,
        "receiver_firmware_version": "1.2.3",
        "serial_buffer_rx_size": 100000,
        "serial_buffer_tx_size": 1280,
        "turbine_id": "unknown"
      },
      "nodes": {
        "0": {
          "acc_freq": 100,
          "acc_range": 16,
          "analog_freq": 16384,
          "baros_bm": 1023,
          "baros_freq": 100,
          "blade_id": "0",
          "constat_period": 45,
          "battery_info_period": 3600,
          "decline_reason": {
            "0": "Bad block detection ongoing",
            "1": "Task already registered, cannot register again",
            "2": "Task is not registered, cannot de-register",
            "3": "Connection Parameter update unfinished"
          },
          "diff_baros_freq": 1000,
          "initial_node_handles": {
            "34": "Abs. baros",
            "36": "Diff. baros",
            "38": "Mic 0",
            "40": "Mic 1",
            "42": "IMU Accel",
            "44": "IMU Gyro",
            "46": "IMU Magnetometer",
            "48": "Analog1",
            "50": "Analog2",
            "52": "Constat",
            "54": "Cmd Decline",
            "56": "Sleep State",
            "58": "Info Message"
          },
          "gyro_freq": 100,
          "gyro_range": 2000,
          "remote_info_type": {
            "0": "Battery info"
          },
          "mag_freq": 12.5,
          "mics_freq": 15625,
          "mics_bm": 1023,
          "max_timestamp_slack": 0.005,
          "max_period_drift": 0.02,
          "node_firmware_version": "unknown",
          "number_of_sensors": {
            "Mics": 10,
            "Baros_P": 40,
            "Baros_T": 40,
            "Diff_Baros": 5,
            "Acc": 3,
            "Gyro": 3,
            "Mag": 3,
            "Analog Vbat": 1,
            "Constat": 4,
            "battery_info": 3
          },
          "periods": {
            "Mics": 6.4e-5,
            "Baros_P": 0.01,
            "Baros_T": 0.01,
            "Diff_Baros": 0.001,
            "Acc": 0.01,
            "Gyro": 0.01,
            "Mag": 0.08,
            "Analog Vbat": 6.103515625e-5,
            "Constat": 0.045,
            "battery_info": 3600
          },
          "samples_per_packet": {
            "Mics": 8,
            "Diff_Baros": 24,
            "Baros_P": 1,
            "Baros_T": 1,
            "Acc": 40,
            "Gyro": 40,
            "Mag": 40,
            "Analog Vbat": 60,
            "Constat": 24,
            "battery_info": 1
          },
          "sensor_conversion_constants": {
            "Mics": [1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
            "Diff_Baros": [1, 1, 1, 1, 1],
            "Acc": [1, 1, 1],
            "Gyro": [1, 1, 1],
            "Mag": [1, 1, 1],
            "Analog Vbat": [1],
            "Constat": [1, 1, 1, 1],
            "battery_info": [1e6, 100, 256]
          },
          "sensor_coordinates": {
            "Mics": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "Baros_P": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "Baros_T": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "Diff_Baros": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "Acc": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "Gyro": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "Mag": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "Analog Vbat": [[0, 0, 0]],
            "Constat": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ],
            "battery_info": [
              [0, 0, 0],
              [0, 0, 0],
              [0, 0, 0]
            ]
          },
          "sensor_names": [
            "Mics",
            "Baros_P",
            "Baros_T",
            "Diff_Baros",
            "Acc",
            "Gyro",
            "Mag",
            "Analog Vbat",
            "Constat",
            "battery_info"
          ],
          "sleep_state": {
            "0": "Exiting sleep",
            "1": "Entering sleep"
          }
        }
      },
      "measurement_campaign": {
        "label": "my-test-1",
        "description": null
      }
    }



data_gateway/configuration.py
import copy

from data_gateway.exceptions import WrongNumberOfSensorCoordinatesError


BASE_STATION_ID = "base-station"

DEFAULT_SENSOR_NAMES = [
    "Mics",
    "Baros_P",
    "Baros_T",
    "Diff_Baros",
    "Acc",
    "Gyro",
    "Mag",
    "Analog Vbat",
    "Constat",
    "battery_info",
]

DEFAULT_INITIAL_GATEWAY_HANDLES = {
    "64": "Local Info Message",
}

DEFAULT_INITIAL_NODE_HANDLES = {
    "34": "Abs. baros",
    "36": "Diff. baros",
    "38": "Mic 0",
    "40": "Mic 1",
    "42": "IMU Accel",
    "44": "IMU Gyro",
    "46": "IMU Magnetometer",
    "48": "Analog1",
    "50": "Analog2",
    "52": "Constat",
    "54": "Cmd Decline",
    "56": "Sleep State",
    "58": "Remote Info Message",
    "60": "Timestamp Packet 0",
    "62": "Timestamp Packet 1",
    "64": "Local Info Message",
}

DEFAULT_DECLINE_REASONS = {
    "0": "Bad block detection ongoing",
    "1": "Task already registered, cannot register again",
    "2": "Task is not registered, cannot de-register",
    "3": "Connection parameter update unfinished",
    "4": "Not ready to sleep",
    "5": "Not in sleep",
}

DEFAULT_SLEEP_STATES = {"0": "Exiting sleep", "1": "Entering sleep"}

DEFAULT_REMOTE_INFO_TYPES = {"0": "Battery info", "1": "Status feedback"}

DEFAULT_LOCAL_INFO_TYPES = {
    "0": "Synchronization not ready as not every sensor node is connected",
    "1": "Time synchronization info",
    "2": "Time sync exception",
    "4": "Time sync coarse data record error",
    "8": "Time sync alignment error",
    "16": "Time sync coarse data time diff error",
    "32": "Device not connected",
    "64": "Select message destination successful",
    "128": "Time sync success",
    "129": "Coarse sync finish",
    "130": "Time sync msg sent",
    "240": "Command not registered",
}

DEFAULT_SAMPLES_PER_PACKET = {
    "Mics": 8,
    "Diff_Baros": 24,
    "Baros_P": 1,
    "Baros_T": 1,
    "Acc": 40,  # IMU, int(240 / 2 / 3)
    "Gyro": 40,  # IMU, int(240 / 2 / 3)
    "Mag": 40,  # IMU, int(240 / 2 / 3)
    "Analog Vbat": 60,
    "Constat": 24,
    "battery_info": 1,
}

DEFAULT_SENSOR_CONVERSION_CONSTANTS = {
    "Mics": 1,
    "Diff_Baros": 1,
    "Baros_P": 40.96,
    "Baros_T": 100,
    "Acc": 1,
    "Gyro": 1,
    "Mag": 1,
    "Analog Vbat": 1,
    "Constat": 1,
    "battery_info": 1,
}

DEFAULT_SENSOR_COMMANDS = {
    "start": ["startBaros", "startDiffBaros", "startIMU", "startMics"],
    "stop": ["stopBaros", "stopDiffBaros", "stopIMU", "stopMics"],
    "configuration": ["configBaros", "configAccel", "configGyro", "configMics"],
    "utilities": [
        "getBattery",
        "setConnInterval",
        "tpcBoostIncrease",
        "tpcBoostDecrease",
        "tpcBoostHeapMemThr1",
        "tpcBoostHeapMemThr2",
        "tpcBoostHeapMemThr4",
    ],
}

DEFAULT_NUMBER_OF_SENSORS = {
    "Mics": 10,
    "Baros_P": 40,
    "Baros_T": 40,
    "Diff_Baros": 5,
    "Acc": 3,
    "Gyro": 3,
    "Mag": 3,
    "Analog Vbat": 2,
    "Constat": 4,
    "battery_info": 3,
}

DEFAULT_MEASUREMENT_CAMPAIGN_METADATA = {
    "label": None,
    "description": None,
}

HANDLE_DEFINITION_PACKET_TYPE = 255  # 0xFF,


class GatewayConfiguration:
    """A data class containing configured/default values for the gateway receiver

    :param float baudrate: serial port baud rate
    :param Literal["little", "big"] endian: one of "little" or "big"
    :param str installation_reference: A unique reference (id) for the current installation
    :param float latitude: The latitude of the turbine in WGS84 coordinate system
    :param dict|None local_info_types: A map of labels to the type of information received from base station
    :param float longitude: The longitude of the turbine in WGS84 coordinate system
    :param int packet_key: The prefix value for packets received from the base station (i.e. not from a node)
    :param int packet_key_offset: The base prefix value for packets received from nodes (node_packet_key = node_id + packet_key_offset)
    :param str receiver_firmware_version: The version of the firmware running on the gateway receiver, if known.
    :param int serial_buffer_rx_size: serial receiving buffer size in bytes
    :param int serial_buffer_tx_size: serial transmitting buffer size in bytes
    :param str turbine_id: A unique id for the turbine on which this is installed
    :return None:
    """

    def __init__(
        self,
        baudrate=2300000,
        endian="little",
        initial_gateway_handles=None,
        installation_reference="unknown",
        latitude=0,
        local_info_types=None,
        longitude=0,
        packet_key=254,
        packet_key_offset=245,
        receiver_firmware_version="unknown",
        serial_buffer_rx_size=4095,
        serial_buffer_tx_size=1280,
        turbine_id="unknown",
    ):
        self.baudrate = baudrate
        self.endian = endian
        self.initial_gateway_handles = initial_gateway_handles or DEFAULT_INITIAL_GATEWAY_HANDLES
        self.installation_reference = installation_reference
        self.latitude = latitude
        self.local_info_types = local_info_types or DEFAULT_LOCAL_INFO_TYPES
        self.longitude = longitude
        self.serial_buffer_rx_size = serial_buffer_rx_size
        self.serial_buffer_tx_size = serial_buffer_tx_size
        self.turbine_id = turbine_id
        self.receiver_firmware_version = receiver_firmware_version
        self.packet_key = packet_key
        self.packet_key_offset = packet_key_offset

    def to_dict(self):
        """Convert the configuration to a dictionary.

        :return dict:
        """
        return vars(self)


class NodeConfiguration:
    """A data class containing configured/default values for a sensor node

    :param float acc_freq: accelerometers sampling frequency
    :param float acc_range: TODO nobody seems to know...
    :param float analog_freq: analog sensors sampling frequency
    :param float baros_bm: TODO nobody seems to know...
    :param float baros_freq: barometers sampling frequency
    :param str blade_id: The id of the blade on which the node is mounted, if known
    :param float constat_period: period of incoming connection statistic parameters in ms
    :param int|float battery_info_period: period of battery information in s
    :param dict|None decline_reason:
    :param float diff_baros_freq: differential barometers sampling frequency
    :param dict|None initial_node_handles: Map of the default handles which a node will use to communicate packet type (the expected contents of packet payload). These are defaults, as they may be altered on the fly by a node.
    :param float gyro_freq: gyrometers sampling frequency
    :param float gyro_range: TODO nobody seems to know...
    :param float mag_freq:
    :param float mics_freq: microphones sampling frequency
    :param float mics_bm: TODO nobody seems to know...
    :param float max_timestamp_slack: TODO   # 5ms
    :param float max_period_drift: TODO   # 2% difference between IMU clock and CPU clock allowed
    :param str node_firmware_version: The verison of the firmware on the node, if known.
    :param dict|None number_of_sensors: A map for each sensor, giving the number of samples expected from that sensor
    :param dict|None remote_info_type: A map of labels to the type of information received from remote node
    :param dict|None samples_per_packet: A map for each sensor, giving the number of samples sent in a packet from that sensor
    :param dict|None sensor_commands:
    :param dict|None sensor_conversion_constants:
    :param dict|None sensor_coordinates:
    :param list|None sensor_names: List of sensors present on the measurement node
    :param dict|None sleep_state:
    :return None:
    """

    def __init__(
        self,
        acc_freq=100,
        acc_range=16,
        analog_freq=16384,
        baros_bm=0x3FF,
        baros_freq=100,
        blade_id="unknown",
        constat_period=45,
        battery_info_period=3600,
        decline_reason=None,
        diff_baros_freq=1000,
        initial_node_handles=None,
        gyro_freq=100,
        gyro_range=2000,
        mag_freq=12.5,
        mics_freq=15625,
        mics_bm=0x3FF,
        max_timestamp_slack=5e-3,
        max_period_drift=0.02,
        node_firmware_version="unknown",
        number_of_sensors=None,
        remote_info_type=None,
        samples_per_packet=None,
        sensor_commands=None,
        sensor_conversion_constants=None,
        sensor_coordinates=None,
        sensor_names=None,
        sleep_state=None,
    ):
        # Set kwargs as attributes directly
        self.acc_freq = acc_freq
        self.acc_range = acc_range
        self.analog_freq = analog_freq
        self.baros_bm = baros_bm
        self.baros_freq = baros_freq
        self.blade_id = blade_id
        self.constat_period = constat_period
        self.battery_info_period = battery_info_period
        self.diff_baros_freq = diff_baros_freq
        self.gyro_freq = gyro_freq
        self.gyro_range = gyro_range
        self.mag_freq = mag_freq
        self.max_timestamp_slack = max_timestamp_slack
        self.max_period_drift = max_period_drift
        self.mics_bm = mics_bm
        self.mics_freq = mics_freq
        self.node_firmware_version = node_firmware_version

        # Set default dictionaries
        self.decline_reason = decline_reason or DEFAULT_DECLINE_REASONS
        self.initial_node_handles = initial_node_handles or DEFAULT_INITIAL_NODE_HANDLES
        self.remote_info_type = remote_info_type or DEFAULT_REMOTE_INFO_TYPES
        self.number_of_sensors = number_of_sensors or DEFAULT_NUMBER_OF_SENSORS
        self.samples_per_packet = samples_per_packet or DEFAULT_SAMPLES_PER_PACKET
        self.sensor_commands = sensor_commands or DEFAULT_SENSOR_COMMANDS
        self.sensor_conversion_constants = sensor_conversion_constants or DEFAULT_SENSOR_CONVERSION_CONSTANTS
        self.sensor_names = sensor_names or DEFAULT_SENSOR_NAMES
        self.sleep_state = sleep_state or DEFAULT_SLEEP_STATES

        # Set calculated defaults
        self.sensor_coordinates = sensor_coordinates or self._get_default_sensor_coordinates()

        for sensor, coordinates in self.sensor_coordinates.items():
            number_of_sensors = self.number_of_sensors[sensor]

            if len(coordinates) != number_of_sensors:
                raise WrongNumberOfSensorCoordinatesError(
                    f"The number of sensors for the {sensor!r} sensor type is {number_of_sensors} but coordinates "
                    f"were only given for {len(coordinates)} sensors. Coordinates must be given for every sensor."
                )

        # Ensure conversion constants are consistent
        self._expand_sensor_conversion_constants()

        # Validate the final configuration
        self._check()

    @property
    def periods(self):
        """Get the period in seconds for each sensor (computed from the sensor frequencies).

        :return dict:
        """
        return {
            "Mics": 1 / self.mics_freq,
            "Baros_P": 1 / self.baros_freq,
            "Baros_T": 1 / self.baros_freq,
            "Diff_Baros": 1 / self.diff_baros_freq,
            "Acc": 1 / self.acc_freq,
            "Gyro": 1 / self.gyro_freq,
            "Mag": 1 / self.mag_freq,
            "Analog Vbat": 1 / self.analog_freq,
            "Constat": self.constat_period / 1000,
            "battery_info": self.battery_info_period,
        }

    def to_dict(self):
        """Serialise the configuration to a dictionary.

        :return dict:
        """
        return {**vars(self), "periods": self.periods}

    def _get_default_sensor_coordinates(self):
        return {
            sensor_name: [(0, 0, 0)] * number_of_sensors
            for sensor_name, number_of_sensors in self.number_of_sensors.items()
        }

    def _check(self):
        """Serialise self to JSON then make sure it matches a schema"""
        # NOT IMPLEMENTED YET
        # See https://github.com/aerosense-ai/data-gateway/issues/18

    def _expand_sensor_conversion_constants(self):
        """Expand the sensor conversion constant to arrays of the correct size

        This means that sensor conversion constants can be given as single values rather than lists.
        Passing a full list of adjusted values is still possible, to calibrate individual samples.
        """
        unconverted = self.sensor_conversion_constants
        converted = dict()
        for key, value in unconverted.items():
            number_of_samples = self.number_of_sensors[key]
            if isinstance(value, (int, float)):
                converted[key] = [value] * number_of_samples
            elif isinstance(value, list):
                if len(value) != number_of_samples:
                    raise ValueError(
                        f"If you give a list of conversion constants for {key}, it must be the same length as the number of samples you expect from that sensor (got length {len(value)}, require length {number_of_samples})"
                    )
                converted[key] = value
            else:
                raise ValueError(f"Unknown sensor conversion constant value {value}")

        self.sensor_conversion_constants = converted


class Configuration:
    """Configuration class for gateway, node and measurement campaign configuration data.

    :param dict|None gateway: A dict of values used to customise the gateway configuration
    :param dict|None nodes: A dict of dicts, keyed by node_id, used to customise the configuration for each node
    :param dict|None measurement_campaign: A dict of metadata about the current measurement campaign of the gateway provided by the user
    :return None:
    """

    def __init__(self, gateway=None, nodes=None, measurement_campaign=None, **kwargs):
        gateway_configuration = gateway or {}
        self.gateway = GatewayConfiguration(**gateway_configuration)

        if len(kwargs) > 0:
            raise ValueError(
                "Properties other than 'gateway', 'nodes' and 'measurement_campaign' passed to Configuration. Are you "
                "using an old-format configuration file?"
            )

        # Set up a single-node default in the absence of any nodes at all.
        self.nodes = {}

        if nodes is None:
            self.nodes["0"] = NodeConfiguration()
        else:
            for node_id, node in nodes.items():
                self.nodes[str(node_id)] = NodeConfiguration(**node)

        # Set up the measurement-campaign-specific data as empty.
        self.measurement_campaign = measurement_campaign or DEFAULT_MEASUREMENT_CAMPAIGN_METADATA

    @classmethod
    def from_dict(cls, dictionary):
        """Construct a configuration from a dictionary. Note that a dictionary for each sub-configurations is required
        - the construction will fail if any are missing.

        :param dict dictionary:
        :return Configuration:
        """
        dictionary = copy.deepcopy(dictionary)

        for node in dictionary["nodes"].values():
            if "periods" in node:
                node.pop("periods")

        return cls(
            gateway=dictionary["gateway"],
            nodes=dictionary["nodes"],
            measurement_campaign=dictionary["measurement_campaign"],
        )

    @property
    def node_ids(self):
        """Get the IDs of the nodes in the current configuration.

        :return list: A list of node ids
        """
        return list(self.nodes)

    def get_leading_byte(self, node_id=None):
        """Get the leading byte for a given node ID or for base station packets by default. Uses the packet key and the
        packet key offset.

        :param int|str node_id: The node ID for which you want the packet key
        :return bytes: The leading byte for packets from the given node or the base station
        """
        if node_id is None:
            return self.gateway.packet_key.to_bytes(1, "little")

        node_packet_key = self.gateway.packet_key_offset + int(node_id)
        return node_packet_key.to_bytes(1, self.gateway.endian)

    @property
    def leading_bytes_map(self):
        """Access a dict that maps leading bytes to node_ids (or the base station id)

        :return dict:
        """
        nodes = {self.get_leading_byte(node_id): node_id for node_id in self.node_ids}

        return {
            self.get_leading_byte(): BASE_STATION_ID,
            **nodes,
        }

    def to_dict(self):
        """Serialise the configuration to a dictionary.

        :return dict:
        """
        return {
            "gateway": self.gateway.to_dict(),
            "nodes": {name: node.to_dict() for name, node in self.nodes.items()},
            "measurement_campaign": self.measurement_campaign,
        }
